File: test/services/dt-async/lib/petstore_async_impl.dart
import 'generated/petstore/api/pet.dart';
import 'generated/petstore/api/pet_store.dart';
import 'generated/petstore/api/petstore/addpet/in.dart' as addpet;
import 'generated/petstore/api/petstore/addpet/out.dart' as addpet_out;
import 'generated/petstore/api/petstore/getpet/in.dart' as getpet;
import 'generated/petstore/api/petstore/getpet/out.dart' as getpet_out;
import 'generated/petstore/api/petstore/listpets/in.dart' as listpets;
import 'generated/petstore/api/petstore/listpets/out.dart' as listpets_out;
import 'generated/petstore/api/petstore/deletepet/in.dart' as deletepet;
import 'generated/petstore/api/petstore/deletepet/out.dart' as deletepet_out;

/// Async server implementation of the generated PetStore interface with
/// Future<T> return types.
///
/// With --dt-async-services=true (fixed in T80/T81), the generated PetStore
/// abstract class declares Future<T> return types on all methods and the
/// server dispatchers use async/await. This impl passes dart analyze and
/// serves as a GREEN regression guard for D25.
class PetStoreAsyncImpl implements PetStore {
  final Map<int, Pet> _pets = {};
  int _nextId = 1;

  void reset() {
    _pets.clear();
    _nextId = 1;
  }

  @override
  Future<addpet_out.out> addPet(addpet.in_ arg) async {
    final petId = _nextId;
    _nextId += 1;
    final pet = Pet(id: petId, name: arg.name, status: arg.status, tag: arg.tag);
    _pets[petId] = pet;
    return addpet_out.out(pet: pet);
  }

  @override
  Future<getpet_out.out> getPet(getpet.in_ arg) async {
    final pet = _pets[arg.id];
    if (pet == null) {
      throw StateError('Pet not found: id=${arg.id}');
    }
    return getpet_out.out(pet: pet);
  }

  @override
  Future<listpets_out.out> listPets(listpets.in_ arg) async {
    final sorted = _pets.values.toList()..sort((a, b) => a.id.compareTo(b.id));
    return listpets_out.out(pets: sorted);
  }

  @override
  Future<deletepet_out.out> deletePet(deletepet.in_ arg) async {
    final existed = _pets.containsKey(arg.id);
    if (existed) {
      _pets.remove(arg.id);
    }
    return deletepet_out.out(deleted: existed);
  }
}

File: test/services/kt-async/src/main/kotlin/example/PetStoreAsyncImpl.kt
package example

import java.util.concurrent.atomic.AtomicLong

/**
 * Async server implementation of the generated PetStore interface using
 * `suspend fun` methods.
 *
 * With --kt-async-services=true (fixed in T76/T77), the generated PetStore
 * interface declares `suspend fun` methods and the server dispatchers
 * `suspend`-call the impl. This impl compiles and round-trips correctly as a
 * GREEN regression guard for D25.
 */
class PetStoreAsyncImpl : petstore.api.PetStore {
    private val pets = mutableMapOf<Long, petstore.api.Pet>()
    private val nextId = AtomicLong(1)

    fun reset() {
        pets.clear()
        nextId.set(1)
    }

    override suspend fun addPet(arg: petstore.api.petstore.addpet.In): petstore.api.petstore.addpet.Out {
        val id = nextId.getAndIncrement()
        val pet = petstore.api.Pet(
            id = id,
            name = arg.name,
            status = arg.status,
            tag = arg.tag
        )
        pets[id] = pet
        return petstore.api.petstore.addpet.Out(pet = pet)
    }

    override suspend fun getPet(arg: petstore.api.petstore.getpet.In): petstore.api.petstore.getpet.Out {
        val pet = pets[arg.id] ?: throw NoSuchElementException("Pet not found: id=${arg.id}")
        return petstore.api.petstore.getpet.Out(pet = pet)
    }

    override suspend fun listPets(arg: petstore.api.petstore.listpets.In): petstore.api.petstore.listpets.Out {
        val sorted = pets.values.sortedBy { it.id }
        return petstore.api.petstore.listpets.Out(pets = sorted)
    }

    override suspend fun deletePet(arg: petstore.api.petstore.deletepet.In): petstore.api.petstore.deletepet.Out {
        val existed = pets.remove(arg.id) != null
        return petstore.api.petstore.deletepet.Out(deleted = existed)
    }
}
